### PR TITLE
Upgrade to React 18

### PR DESCRIPTION
**Pull request by TheReshma**
### Upgrade to React 18

-  Upgraded `react` & `react-dom` to `^18.0.0`
- Upgraded `@types/react` & `@types/react-dom` to `^18.0.0`

All Changes are done as per this [Blog](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html)
Dependency PR for [#21884](https://github.com/vector-im/element-web/pull/21884)
Fixes [#21594](https://github.com/vector-im/element-web/issues/21594)

Signed-off-by: Reshma <reshmashaik3011@gmail.com>

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

**Review by turt2live — COMMENTED** [review]
clearing code review because github won't let us remove it

File: package.json [hunk 1]
```diff
@@ -152,9 +152,9 @@
     "@types/pako": "^1.0.1",
     "@types/parse5": "^6.0.0",
     "@types/qrcode": "^1.3.5",
-    "@types/react": "17.0.14",
+    "@types/react": "^18.0.0",
     "@types/react-beautiful-dnd": "^13.0.0",
-    "@types/react-dom": "17.0.9",
+    "@types/react-dom": "^18.0.0",
     "@types/react-test-renderer": "^17.0.1",
```
Comment (t3chguy) [inline package.json:158]: This probably needs updating too

We also need a react 18 enzyme adapter, see L164 below

Right now you have a conflicting version of react-dom being imported by something

```


node_modules/@types/react-dom/node_modules/@types/react/index.d.ts:3057:14 - error TS2300: Duplicate identifier 'LibraryManagedAttributes'.
--
  |  
  | 3057         type LibraryManagedAttributes<C, P> = C extends React.MemoExoticComponent<infer T> \| React.LazyExoticComponent<infer T>
  | ~~~~~~~~~~~~~~~~~~~~~~~~
  |  
  | node_modules/@types/react/index.d.ts:3127:14
  | 3127         type LibraryManagedAttributes<C, P> = C extends React.MemoExoticComponent<infer T> \| React.LazyExoticComponent<infer T>
  | ~~~~~~~~~~~~~~~~~~~~~~~~
  | 'LibraryManagedAttributes' was also declared here.

Comment (TheReshma) [reply package.json]: On L164 :   `"@wojtekmaj/enzyme-adapter-react-17": "^0.6.1"`
The Author of this package has published a [Blog](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl) saying that there won't be an enzyme adapter for React 18. I searched for other enzyme adapters as well but couldn't find any.

The `error TS2300: Duplicate identifier 'LibraryManagedAttributes'` would probably be resolved now after I updated `"resolutions": {
    "@types/react": "^18.0.0"
  }`  as per this stack overflow [answer](https://stackoverflow.com/questions/52399839/duplicate-identifier-librarymanagedattributes). 

Also updated `"@types/react-test-renderer"` which had react 17 in its dependencies.

Comment (t3chguy) [reply package.json]: > The Author of this package has published a [Blog](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl) saying that there won't be an enzyme adapter for React 18. I searched for other enzyme adapters as well but couldn't find any.

Then we would need to move away from Enzyme to be able to upgrade to React 18?

**Review by t3chguy — COMMENTED** [review]
Looks like React has changed the default type of `context` to be unknown so will need to be set appropriately in all class components which make use of `contextType` as well as other similar errors

**Comment by TheReshma** [comment]
> Looks like React has changed the default type of `context` to be unknown so will need to be set appropriately in all class components which make use of `contextType` as well as other similar errors

Yeah this is used quite often to set the `context` : 
```
import React, { ContextType } from 'react';
public static contextType = MatrixClientContext;
context: ContextType<typeof MatrixClientContext>; 
```

What do I do now ?

**Comment by t3chguy** [comment]
`context: ContextType<typeof MatrixClientContext>;` should be the bit required - https://blog.logrocket.com/upgrading-react-18-typescript/ uses a codemod to add `context: any` everywhere but lets please be more explicit

**Comment by TheReshma** [comment]
Maybe use the **codemod** to remove ` ◉ deprecated-react-type  ◉ deprecated-sfc-element  ◉ deprecated-sfc  ◉ deprecated-stateless-component` & add `implicit-children` , `useCallback-implicit-any`.
But add `context: ContextType<typeof MatrixClientContext>;` **manually** to every class component

**Comment by t3chguy** [comment]
Feel free to use the aid of codemods, I will be reviewing the final diff by hand and pointing out any tweaks we may need to make

**Comment by TheReshma** [comment]
Sure. Thanks !

File: package.json (hunk 1 above)
Comment (t3chguy) [reply package.json]: It is the longer-term plan to move away from Enzyme to RTL but it'd suck if we have to block the 18 upgrade on it

**Comment by t3chguy** [comment]
https://github.com/matrix-org/matrix-react-sdk/runs/6401736836?check_suite_focus=true#step:4:345 might be caused by the broken yarn.lock in the other layer

**Review by t3chguy — CHANGES_REQUESTED** [review]
React.FC components which make use of `children` now need it defining in their prop types explicitly https://twitter.com/reactjs/status/1512453230504124420

```
src/components/views/spaces/SpaceCreateMenu.tsx:154:5 - error TS2339: Property 'children' does not exist on type 'ISpaceCreateFormProps'.

154     children,

**Pull request closed without merging**